compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/cyclefind/TarjanSccSolverAdjacencyList.java

```diff
@@ -0,0 +1,161 @@
+    private List<List<Integer>> adjacencyList;
+    private boolean solved;
+    private int sccCount;
+    private int id;
+    private boolean[] onStack;
+    private int[] ids;
+    private int[] lowLinks;
+    private Deque<Integer> stack;
+    public final List<Integer> dependencyOrder;
+
+    private static final int UNVISITED = -1;
+
+    private TarjanSccSolverAdjacencyList(List<List<Integer>> adjacencyList) {
+        if (adjacencyList == null) {
+            throw new IllegalArgumentException("Graph cannot be null.");
+        }
+        n = adjacencyList.size();
+        this.adjacencyList = adjacencyList;
+        dependencyOrder = new ArrayList<>();
+    }
+
+    /**
+     * Find number of strongly connected components in the graph.
+     * @return number of SCCs
+     */
+    public int sccCount() {
+        if (!solved) {
+            solve();
+        }
+        return sccCount;
+    }
+
+    private void solve() {
+        if (solved) {
+            return;
+        }
+
+        ids = new int[n];
+        lowLinks = new int[n];
+        onStack = new boolean[n];
+        stack = new ArrayDeque<>();
+        Arrays.fill(ids, UNVISITED);
+
+        for (int i = 0; i < n; i++) {
+            if (ids[i] == UNVISITED) {
+                dfs(i);
+            }
+        }
+
+        solved = true;
+    }
+
+    private void dfs(int at) {
+        stack.push(at);
+        onStack[at] = true;
+        ids[at] = lowLinks[at] = id++;
+
+        for (int to : adjacencyList.get(at)) {
+            if (ids[to] == UNVISITED) {
+                dfs(to);
+            }
+            if (onStack[to]) {
+                lowLinks[at] = Math.min(lowLinks[at], lowLinks[to]);
+            }
+        }
+
+        // On recursive callback, if we're at the root node (start of SCC)
+        // empty the seen stack until back to root.
+        if (ids[at] == lowLinks[at]) {
+            for (int node = stack.pop(); true; node = stack.pop()) {
+                onStack[node] = false;
+                lowLinks[node] = ids[at];
+                if (node == at) {
+                    break;
+                }
+            }
+            sccCount++;
+        }
+
+        dependencyOrder.add(at);
+    }
+
+    /**
+     * Initialized adjacency list for graph with n nodes.
+     * @param n number of nodes
+     * @return initialized list to fill with edges
+     */
+    public static TarjanSccSolverAdjacencyList createGraph(int n) {
+        List<List<Integer>> graph = new ArrayList<>(n);
+        for (int i = 0; i < n; i++) {
+            graph.add(new ArrayList<>());
+        }
+        return new TarjanSccSolverAdjacencyList(graph);
+    }
+
+    /**
+     * Adds a directed edge from node 'from' to node 'to'.
+     * @param from edge starting from node
+     * @param to to node
+     * @return graph with added edge
+     */
+    public TarjanSccSolverAdjacencyList addEdge(int from, int to) {
+        this.adjacencyList.get(from).add(to);
+        return this;
```

Comment:
Do we need to return from this method?

Reply:
This was to allow chaining such as addEdge(1,2).addEdge(2,1).addEdge..., I guess this API will not be used like that, will remove.